docs/architecture/adr-002-lite-client.md

```diff
@@ -0,0 +1,439 @@
+mechanism may be utilized.
+
+### Verifier
+
+Most of the lite client logic resides in the verifier, where commits for a
+header are actually verified. In order to abstract over all the data in the
+Tendermint data structures, we use a set of traits that include only the
+information absolutely necessary for the lite client. From this perspective, we
+have the following traits.
+
+#### Header
+
+A Header has only a height, time, and validator sets, and can be identified by its hash:
+
+```rust
+pub trait Header {
+    fn height(&self) -> Height;
+    fn bft_time(&self) -> Time;
+    fn validators_hash(&self) -> Hash;
+    fn next_validators_hash(&self) -> Hash;
+
+    fn hash(&self) -> Hash;
+}
+```
+
+#### Validator
+
+A validator contributes a positive voting power if a message was correctly signed by it,
+otherwise it contributes 0. We could represent this with a single method that
+returns either 0 or the voting power, but it's probably clearer with two methods:
+
+```rust
+pub trait Validator {
+    fn power(&self) -> u64;
+    fn verify_signature(&self, sign_bytes: &[u8], signature: &[u8]) -> bool;
+}
+```
+
+#### Vote
+
+A vote corresponds to the message signed by a particular validator for a particular commit.
+Since it is expected to be for a particular commit, it does not need to include
+that information explicitly (eg. the height, the round, the block ID, etc.).
+Instead, from the verifier's perspective, all that information can be contained implicitly within the message
+bytes that are being signed. When those message bytes and the signature are passed to 
+`validator.verify_signature` for the corresponding validator, it should return true:
+
+```rust
+pub trait Vote {
+    fn validator_id(&self) -> ValID;
+    fn sign_bytes(&self) -> &[u8];
+    fn signature(&self) -> &[u8];
+}
+```
+
+#### Validator Set
+
+A validator set is a collection of validators that we will want to iterate over
+to check if each validator signed their vote correctly.
+We also need to know its hash, so we can check it against what's in a header. 
+We use an associated type for the validator:
+
+```rust
+pub trait ValidatorSet {
+    type Validator: Validator;
+
+    fn hash(&self) -> Hash;
+    fn into_vec(&self) -> Vec<Self::Validator>;
+}
+```
+
+It may seem more natural to use an `iter` method here than `into_vec`, since
+ultimately we only need to iterate over the underlying associated type.
+However, this turns out to be much more difficult than expected due to the lack
+of size information in the `Iter` trait. It's possible to use another associated
+type, eg. `type ValidatorIter: ExactSizeIterator<Item = Self::Validator>;`, but it just
+complicates things significantly. If this becomes a performance issue (ie. we
+end up having to copy the vector of validators), we can address it then.
+
+#### Commit
+
+A commit is a collection of votes that we will want to iterate over so we can
+verify each vote.
```

Comment:
Can we make the API s.t. we can upgrade to aggregate signatures w/o an API change in the lite client?

Essentially make one function that goes from validator index to boolean representing "has signed", and one aggregate signature struct. Verify checks that 2/3rds of stake has signed, and that aggregate signature.verify(<signed pubkeys>) verifies.

Reply:
I thot about this a bit. I feel it might be a bit too early for that. This thing is already pretty abstract and I'm not sure it will pay yet to make it even more so.

Also, come to think of it, will aggregate sigs work with the skipping method? If I have an aggregate signature, can I check if +1/3 of some other validator set (not necessarily the validator set that produced the sig) signed it?

Reply:
You have a better impression of this than I do, but it feels to me like aggregate signatures shouldn't be too far out? Related projects are already using them (LibraBFT, Nervos (https://github.com/cryptape/overlord/) a TM fork)

To check an aggregate signature, you must have all the signing public keys. Is this 1/3rd check needed for `checkSupport`? If so, I believe that it is fine, since you are provided the signing public keys in order to verify.

Reply:
Ok so a number of things have come to my attention that suggest this abstraction isn't good enough:

1) Doesn't support aggregate sigs
2) Doesn't support batch verification (whoops!)
3) Testing is actually not much easier than using the actual Tendermint data types.

So all this leads me to believe that the commit trait should be abstracted further to not expose individual signatures but only information about whether say X% signed. In the interest of making some progress in the short term though I think we'll move forward as is and open an issue to deal with this at a later date.

As for the +1/3 check, we have the following situation:
- I trust validator set V
- I get a commit C from validator set V', where V != V'
- I want to check if +1/3 of V is in V' and signed C 
- the +1/3 from V may comprise an arbitrary percentage of V' (ie. they could be 1% or 99%, it doesn't matter). 

Can this be done for an aggregate sig?

Reply:
Yes this can be done. 
The aggregate signature will contain a list of who signed it, so you have the same situation as the non aggregate case. (The verifier has a set of signers, and has to check if 1/3 of V is in this set)